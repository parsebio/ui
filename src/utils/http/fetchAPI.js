import APIError from 'utils/errors/http/APIError';
import FetchError from 'utils/errors/http/FetchError';
import getApiEndpoint from 'utils/apiEndpoint';
import getAuthJWT from 'utils/getAuthJWT';

const fetchAPI = async (path, params = {}, extras = {}) => {
  const headers = params.headers ? params.headers : {};
  const authJWT = extras.jwt || await getAuthJWT();
  const parseJson = extras.parseJson ?? 'true';

  const parameters = {
    ...params,
    headers: {
      ...headers,
      ...(authJWT && { Authorization: `Bearer ${authJWT}` }),
    },
  };

  const url = getApiEndpoint(extras.uiUrl) + path;

  let response;
  try {
    response = await fetch(url, parameters);
  } catch (e) {
    // wrap fetch errors in custom error
    throw new FetchError(e);
  }
  if (!response.ok) {
    let data;
    try {
      data = await response.json();
    } catch (e) {
      // if we can't get extra error info from the response we don't want to fail
      // just return the error code, this happens in many tests
      // where we mock a string response instead of proper json
    }
    // data.message & data.errors follow error formatting defined in:
    // HTTPError.v1.yaml
    throw new APIError(response.status, data?.message, data?.errors);
  }

  if (!parseJson) return response;

  return await response.json();
};

export default fetchAPI;
