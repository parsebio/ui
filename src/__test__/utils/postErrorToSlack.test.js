import postErrorToSlack from 'utils/postErrorToSlack';
import fetchAPI from 'utils/http/fetchAPI';

jest.mock('utils/http/fetchAPI');

const mockError = { message: 'mockMessage', stack: 'Mock error stack' };
const mockReduxDump = {
  cellInfo: {
    focus: {
      store: 'cellSets',
      key: 'louvain',
    },
    groupedTrack: 'louvain',
    selectedTracks: [
      'louvain',
    ],
  },
  cellSets: {
    properies: {
      'louvain-0': {
        cellIds: new Set(Array(25).fill(0)),
      },
    },
  },
  cellMeta: {
    mitochondrialContent: {
      loading: true,
      error: false,
      data: new Array(25).fill(0),
    },
  },
  networkResources: {
    environment: 'test',
  },
};

jest.useFakeTimers('modern').setSystemTime(new Date('2020-01-01').getTime());

delete window.location;
window.location = {
  href: 'http://localhost:3000/experiments/testae48e318dab9a1bd0bexperiment/data-exploration',
};

jest.mock('@aws-amplify/auth', () => ({
  Auth: {
    currentAuthenticatedUser: () => Promise.resolve({
      attributes: {
        name: 'John Doe',
        email: 'fake@email.com',
      },
      username: '5152fake-eb52-474c-user-mocke8c8user',
    }),
  },
}));

jest.mock('stacktrace-js', () => ({
  fromError: () => Promise.resolve(['line1', 'line2']),
}));

describe('PostErrorToSlack', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('Posts requests correctly', async () => {
    await postErrorToSlack(mockError, mockReduxDump);

    expect(fetchAPI).toHaveBeenCalledTimes(1);
    expect(fetchAPI.mock.calls[0]).toMatchSnapshot();

    const { body: formData } = fetchAPI.mock.calls[0][1];
    expect(formData).toMatchSnapshot();
  });

  it('Should not throw an error if POSTing fails', async () => {
    fetchAPI.mockImplementation(() => Promise.reject(new Error('Failed')));

    await postErrorToSlack(mockError, mockReduxDump);

    expect(fetchAPI).toHaveBeenCalledTimes(1);
  });
});
