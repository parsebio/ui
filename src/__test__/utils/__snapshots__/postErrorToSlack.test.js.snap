// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostErrorToSlack Posts requests correctly 1`] = `
[
  "/v2/sendSlackMessage",
  {
    "body": "{"data":{"title":"UI Error at 2020-01-01T00:00:00.000Z","initial_comment":"\\n  ⚠  [TEST] Uncaught UI Error - ExpID testae48e318dab9a1bd0bexperiment - 2020-01-01T00:00:00.000Z\\n  URL: http://localhost:3000/experiments/testae48e318dab9a1bd0bexperiment/data-exploration\\n\\n  User: John Doe <fake@email.com> 5152fake-eb52-474c-user-mocke8c8user\\n  Experiment ID: testae48e318dab9a1bd0bexperiment","channels":"error-logs","filename":"UI-error-testae48e318dab9a1bd0bexperiment-2020-01-01T00:00:00.000Z.log","content":"\\n    ===== ERROR STACK =====\\n    Message:\\n    mockMessage\\n    At:\\n    line1\\nline2\\n\\n    === DETAILS ===\\n    User: John Doe <fake@email.com> 5152fake-eb52-474c-user-mocke8c8user\\n    ExperimentID: testae48e318dab9a1bd0bexperiment\\n    URL: http://localhost:3000/experiments/testae48e318dab9a1bd0bexperiment/data-exploration\\n    Timestamp: 2020-01-01T00:00:00.000Z\\n\\n    ===== REDUX STATE =====\\n    {\\n  \\"cellInfo\\": {\\n    \\"focus\\": {\\n      \\"store\\": \\"cellSets\\",\\n      \\"key\\": \\"louvain\\"\\n    },\\n    \\"groupedTrack\\": \\"louvain\\",\\n    \\"selectedTracks\\": [\\n      \\"louvain\\"\\n    ]\\n  },\\n  \\"cellSets\\": {\\n    \\"properies\\": {\\n      \\"louvain-0\\": {\\n        \\"cellIds\\": {}\\n      }\\n    }\\n  },\\n  \\"cellMeta\\": {\\n    \\"mitochondrialContent\\": {\\n      \\"loading\\": true,\\n      \\"error\\": false,\\n      \\"data\\": [\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        \\"...\\"\\n      ]\\n    }\\n  },\\n  \\"networkResources\\": {\\n    \\"environment\\": \\"test\\"\\n  }\\n}"},"type":"file"}",
    "headers": {
      "Content-Type": "application/json",
    },
    "method": "POST",
  },
]
`;

exports[`PostErrorToSlack Posts requests correctly 2`] = `"{"data":{"title":"UI Error at 2020-01-01T00:00:00.000Z","initial_comment":"\\n  ⚠  [TEST] Uncaught UI Error - ExpID testae48e318dab9a1bd0bexperiment - 2020-01-01T00:00:00.000Z\\n  URL: http://localhost:3000/experiments/testae48e318dab9a1bd0bexperiment/data-exploration\\n\\n  User: John Doe <fake@email.com> 5152fake-eb52-474c-user-mocke8c8user\\n  Experiment ID: testae48e318dab9a1bd0bexperiment","channels":"error-logs","filename":"UI-error-testae48e318dab9a1bd0bexperiment-2020-01-01T00:00:00.000Z.log","content":"\\n    ===== ERROR STACK =====\\n    Message:\\n    mockMessage\\n    At:\\n    line1\\nline2\\n\\n    === DETAILS ===\\n    User: John Doe <fake@email.com> 5152fake-eb52-474c-user-mocke8c8user\\n    ExperimentID: testae48e318dab9a1bd0bexperiment\\n    URL: http://localhost:3000/experiments/testae48e318dab9a1bd0bexperiment/data-exploration\\n    Timestamp: 2020-01-01T00:00:00.000Z\\n\\n    ===== REDUX STATE =====\\n    {\\n  \\"cellInfo\\": {\\n    \\"focus\\": {\\n      \\"store\\": \\"cellSets\\",\\n      \\"key\\": \\"louvain\\"\\n    },\\n    \\"groupedTrack\\": \\"louvain\\",\\n    \\"selectedTracks\\": [\\n      \\"louvain\\"\\n    ]\\n  },\\n  \\"cellSets\\": {\\n    \\"properies\\": {\\n      \\"louvain-0\\": {\\n        \\"cellIds\\": {}\\n      }\\n    }\\n  },\\n  \\"cellMeta\\": {\\n    \\"mitochondrialContent\\": {\\n      \\"loading\\": true,\\n      \\"error\\": false,\\n      \\"data\\": [\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        0,\\n        \\"...\\"\\n      ]\\n    }\\n  },\\n  \\"networkResources\\": {\\n    \\"environment\\": \\"test\\"\\n  }\\n}"},"type":"file"}"`;
